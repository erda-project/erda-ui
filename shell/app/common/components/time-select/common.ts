// Copyright (c) 2021 Terminus, Inc.
//
// This program is free software: you can use, redistribute, and/or modify
// it under the terms of the GNU Affero General Public License, version 3
// or later ("AGPL"), as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

import i18n from 'i18n';
import moment, { Moment } from 'moment';

export type IRelativeTime = keyof typeof relativeTimeRange;

export type IRefreshDuration = keyof typeof autoRefreshDuration;

export interface ITimeRange {
  mode: 'quick' | 'customize';
  quick?: IRelativeTime;
  customize: {
    start?: Moment;
    end?: Moment;
  };
}

export const defaultFormat = 'YYYY-MM-DD HH:mm:ss';

export const relativeTimeRange = {
  'minutes:15': i18n.t('last {amount} {unit}', { amount: 15, unit: i18n.t('common:minutes') }),
  'minutes:30': i18n.t('last {amount} {unit}', { amount: 30, unit: i18n.t('common:minutes') }),
  'hours:1': i18n.t('last {amount} {unit}', { amount: 1, unit: i18n.t('hours') }),
  'hours:3': i18n.t('last {amount} {unit}', { amount: 3, unit: i18n.t('hours') }),
  'hours:6': i18n.t('last {amount} {unit}', { amount: 6, unit: i18n.t('hours') }),
  'hours:12': i18n.t('last {amount} {unit}', { amount: 12, unit: i18n.t('hours') }),
  'days:1': i18n.t('last {amount} {unit}', { amount: 1, unit: i18n.t('days') }),
  'days:3': i18n.t('last {amount} {unit}', { amount: 3, unit: i18n.t('days') }),
  'weeks:1': i18n.t('last {amount} {unit}', { amount: 1, unit: i18n.t('weeks') }),
  'months:1': i18n.t('last {amount} {unit}', { amount: 1, unit: i18n.t('months') }),
  'months:3': i18n.t('last {amount} {unit}', { amount: 3, unit: i18n.t('months') }),
  'months:6': i18n.t('last {amount} {unit}', { amount: 6, unit: i18n.t('months') }),
  today: i18n.t('today'),
  yesterday: i18n.t('yesterday'),
  currentWeek: i18n.t('current week'),
  lastWeek: i18n.t('last week'),
  currentMonth: i18n.t('current month'),
  lastMonth: i18n.t('last month'),
};

export const autoRefreshDuration = {
  'seconds:5': '5s',
  'seconds:10': '10s',
  'seconds:30': '30s',
  'minutes:1': '1m',
  'minutes:5': '5m',
  'minutes:30': '30m',
  'hours:1': '1h',
  'hours:2': '2h',
  'days:1': '1d',
};

/**
 * @description convert relative time range to absolute time range
 * @param unit
 * @param count
 * @returns {[Moment, Moment]}
 */
export const translateRelativeTime = (unit: string, count?: number) => {
  let start = moment();
  let end = start.clone();
  switch (unit) {
    case 'minutes':
    case 'hours':
    case 'days':
    case 'weeks':
    case 'months':
      start = start.subtract(count, unit);
      break;
    case 'today':
      start = start.startOf('day');
      break;
    case 'yesterday':
      start = start.startOf('day').subtract(1, 'days');
      end = end.endOf('day').subtract(1, 'days');
      break;
    case 'currentWeek':
      start = start.startOf('week');
      break;
    case 'lastMonth':
      start = start.startOf('month').subtract(1, 'months');
      end = end.subtract(1, 'months').endOf('month');
      break;
    case 'lastWeek':
      start = start.startOf('week').subtract(1, 'weeks');
      end = end.endOf('week').subtract(1, 'weeks');
      break;
    case 'currentMonth':
      start = start.startOf('month');
      break;
    default:
      start = start.startOf('day');
      break;
  }
  return [start, end];
};

/**
 * @description calculate the auto refresh duration
 * @param count
 * @param unit
 * @return {number}
 */
export const translateAutoRefreshDuration = (count: number, unit: string) => {
  const map = {
    seconds: 1000,
    minutes: 60 * 1000,
    hours: 60 * 60 * 1000,
    days: 24 * 60 * 60 * 1000,
  };
  return count * map[unit];
};

export const transformRange = (v: ITimeRange, format = defaultFormat) => {
  let dateStr;
  let dateArr: Moment[] = [];
  if (v.mode === 'quick' && v.quick) {
    dateStr = relativeTimeRange[v.quick];
    const [unit, count] = v.quick?.split(':') || [];
    if (unit) {
      dateArr = translateRelativeTime(unit, parseInt(count, 10));
    }
  } else if (v.mode === 'customize') {
    const { start, end } = v.customize;
    if (start && end) {
      dateStr = `${start.format(format)} - ${end.format(format)}`;
      dateArr = [start, end];
    }
  }
  return {
    date: dateArr,
    dateStr,
  };
};
