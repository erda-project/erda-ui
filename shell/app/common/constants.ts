// Copyright (c) 2021 Terminus, Inc.
//
// This program is free software: you can use, redistribute, and/or modify
// it under the terms of the GNU Affero General Public License, version 3
// or later ("AGPL"), as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

export const WORKSPACE_LIST = ['DEV', 'TEST', 'STAGING', 'PROD'];
export const DOC_DOMAIN = 'dice-docs.app.terminus.io';
export const FULL_DOC_DOMAIN = `https://${DOC_DOMAIN}`;
export const HELP_DOCUMENT = `${FULL_DOC_DOMAIN}/${process.env.mainVersion}/manual/deploy/resource-management.html#%E7%AE%A1%E7%90%86%E9%85%8D%E9%A2%9D`;
