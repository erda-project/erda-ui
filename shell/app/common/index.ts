// Copyright (c) 2021 Terminus, Inc.
//
// This program is free software: you can use, redistribute, and/or modify
// it under the terms of the GNU Affero General Public License, version 3
// or later ("AGPL"), as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

export { connectCube } from './components/connect';
export { Icon } from './components/icon';
export { ErdaIcon, ErdaCustomIcon } from './components/erda-icon/erda-icon';
export { Panel } from './components/panel';
export { ImgHolder } from './components/img-holder';
export { AddMemberModal } from './components/add-member-modal';
export { FormModal } from './components/form-modal';
export { ImageUpload } from './components/image-upload';
export { Terminal } from './components/terminal';
export { default as CommonRangePicker } from './components/common-range-picker';
export { CRUDTable, CRUDStoreTable } from './components/crud-table';
export { createCRUDStore, createCRUDService } from './stores/_crud_module';
export { RenderForm, RenderPureForm, ReadonlyForm } from './components/render-form';
export { RenderFormItem } from './components/render-formItem';
export { Menu } from './components/tab-menu';
export { DeleteConfirm } from './components/delete-confirm';
export { KeyValueTable } from './components/key-value-table';
export { KeyValueTextArea } from './components/key-value-textarea';
export { KeyValueEditor } from './components/key-value-editor';
export { KeyValueList } from './components/key-value-list';
export { FilterSelect } from './components/filter-select';
export { MembersTable } from './components/members-table';
export { LoadMore } from './components/load-more';
export { Responsive } from './components/layout';
export { CircleProgress } from './components/circle-progress';
export { Copy } from './components/copy';
export { JsonChecker } from './components/json-checker';
export { PagingTable } from './components/paging-table';
export { LoadMoreSelector } from './components/load-more-selector';
export { MemberSelector, AddMemberSelector } from './components/member-selector';
export { ErrorBoundary, errorCatcher } from './components/error-boundary';
export { Holder, EmptyHolder, EmptyListHolder } from './components/empty-holder';
export { LazyRender } from './components/lazy-render';
export { DebounceSearch } from './components/debounce-search';
export { SearchTable } from './components/search-table';
export { OperationBar } from './components/operation-bar';
export { SettingsTabs } from './components/setting-tabs';
export { LogRoller as PureLogRoller } from './components/log/log-roller';
export { CompSwitcher } from './components/comp-switcher';
export { ConfirmDelete } from './components/confirm-delete';
export { Avatar, AvatarList } from './components/avatar';

export { IF } from './components/statement';
export { NoAuthTip } from './components/auth';
export { ConfigLayout } from './components/config-layout';
export { MenuPopover } from './components/menu-popover';
export { FileEditor } from './components/file-editor';
export { CardsLayout } from './components/cards-layout';
export { SwitchAutoScroll } from './components/switch-auto-scroll';
export { default as FileSelect } from './components/file-select';
export { EditField } from './components/edit-field';
export { default as MarkdownEditor } from './components/markdown-editor';
export { default as ColorLabel } from './components/color-label';
export { DropdownSelect } from './components/dropdown-select';
export {
  FilterGroup,
  FilterGroupDrawer,
  FilterGroupV,
  ToolBarWithFilter,
  FilterBarHandle,
} from './components/filter-group/filter-group';
export { default as TimeSelector } from './components/monitor/timeSelector';
export { default as TimeSelect } from './components/time-select/time-select';
export { default as SimpleLog } from './components/runtime/simple-log';
export { default as SimpleLogRoller } from './components/runtime/simple-log-roller';
export { ProtocolSelector, ProtocolInput } from './components/protocol-selector';
export { KVPair } from './components/kv-pair/kv-pair';
export { CompactSelect } from './components/compact-select';
export { Filter, PureFilter } from './components/filter';
// export type { IDiceFilterProps, IPureFilterProps } from './components/filter';
export { ContractiveFilter } from './components/contractive-filter';
export * from './components/use-hooks';
export { default as LogRoller } from './containers/log-roller';
export { default as MonitorChart } from './containers/monitor-chart';
export { default as MetricsMonitor } from './containers/monitor-chart-panel';
export { useFilter, useMultiFilter, CustomFilter } from './components/custom-filter';
export { TagsRow } from './components/tags-row';
export { BoardGrid, PureBoardGrid } from './components/board-grid';
export { LoadingSkeleton, LoadingContent } from './components/loading-skeleton';
export { default as BackToTop } from './components/back-to-top';
export { default as TableActions } from './components/operate-btn';
export { default as UserInfo } from './components/user-info';
export { default as DetailsPanel } from './components/details-panel';
export { default as MultiInput } from './components/multi-input';
export { default as MultiSelect } from './components/multi-select';
export { TreeCategory } from './components/tree/tree';
export { default as EditList } from './components/edit-list/edit-list';
export { default as InputSelect } from './components/input-select/input-select';
export { SortDragGroupList } from './components/sort-drag-list';
