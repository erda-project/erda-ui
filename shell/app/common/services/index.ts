// Copyright (c) 2021 Terminus, Inc.
//
// This program is free software: you can use, redistribute, and/or modify
// it under the terms of the GNU Affero General Public License, version 3
// or later ("AGPL"), as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

import agent from 'agent';

export const fetchLog = ({
  fetchApi,
  ...rest
}: {
  [k: string]: any;
  fetchApi?: string;
}): { lines: COMMON.LogItem[] } => {
  return agent
    .get(fetchApi || '/api/runtime/logs')
    .query(rest)
    .then((response: any) => response.body);
};

export function getMembers(payload: MEMBER.GetListServiceQuery) {
  return agent
    .get('/api/members')
    .query(payload)
    .then((response: any) => response.body);
}

export function getRoleMap(payload: MEMBER.GetRoleTypeQuery): IPagingResp<MEMBER.IRoleType> {
  return agent
    .get('/api/members/actions/list-roles')
    .query(payload)
    .then((response: any) => response.body);
}

export function updateMembers(payload: MEMBER.UpdateMemberBody) {
  return agent
    .post('/api/members')
    .send({ ...payload, options: { rewrite: true } }) // 接口上写死options.rewrite=true，避免新增用户（已有的用户）设置角色无用
    .then((response: any) => response.body);
}

export function removeMember(payload: MEMBER.RemoveMemberBody) {
  return agent
    .post('/api/members/actions/remove')
    .send(payload)
    .then((response: any) => response.body);
}

export const getMemberLabels = (): { list: Array<{ label: string; name: string }> } => {
  return agent.get('/api/members/actions/list-labels').then((response: any) => response.body);
};

export const getUsers = (payload: { q?: string; userID?: string | string[] }) => {
  return agent
    .get('/api/users')
    .query(payload)
    .then((response: any) => response.body);
};

export const getUsersNew = (payload: Merge<IPagingReq, { q?: string }>) => {
  return agent
    .get('/api/users/actions/search')
    .query(payload)
    .then((response: any) => response.body);
};

export const getApps = ({
  pageSize,
  pageNo,
  projectId,
  q,
  searchKey,
  memberID,
  ...rest
}: APPLICATION.GetAppList): IPagingResp<IApplication> => {
  return agent
    .get('/api/applications')
    .query({ pageSize, pageNo, projectId, q: q || searchKey, memberID, ...rest })
    .then((response: any) => response.body);
};

export const uploadFile = (file: any): IUploadFile => {
  return agent
    .post('/api/files')
    .send(file)
    .then((response: any) => response.body);
};

export const genOrgInviteCode = (payload: { orgId: number }): { verifyCode: string } => {
  return agent
    .post('/api/orgs/actions/gen-verify-code')
    .send(payload)
    .then((response: any) => response.body);
};

export const getRenderPageLayout = (payload: CONFIG_PAGE.RenderConfig) => {
  return agent
    .post(`/api/component-protocol/actions/render?scenario=${payload.scenario.scenarioKey}`)
    .send(payload)
    .then((response: any) => response.body);
};
