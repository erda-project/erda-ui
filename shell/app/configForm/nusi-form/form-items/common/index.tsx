// Copyright (c) 2021 Terminus, Inc.
//
// This program is free software: you can use, redistribute, and/or modify
// it under the terms of the GNU Affero General Public License, version 3
// or later ("AGPL"), as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

import React from 'react';
import { Tooltip } from 'app/nusi';
import { Help as IconHelp } from '@icon-park/react';

export { createCombiner } from './combiner';

export const getLabel = (label: string, labelTip: string) => {
  let _label: any = label;
  if (labelTip) {
    _label = (
      <span>
        {label}&nbsp;
        <Tooltip title={labelTip}>
          <IconHelp className="text-icon" />
        </Tooltip>
      </span>
    );
  }
  return _label;
};

export const noop = (a: any) => a;
