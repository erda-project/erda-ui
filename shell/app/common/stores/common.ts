// Copyright (c) 2021 Terminus, Inc.
//
// This program is free software: you can use, redistribute, and/or modify
// it under the terms of the GNU Affero General Public License, version 3
// or later ("AGPL"), as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

import { createStore } from 'core/cube';
import { uploadFile, fetchLog, getRenderPageLayout } from '../services';
import { has } from 'lodash';

interface State {
  logsMap: Obj<COMMON.LOG>;
  slidePanelComps: COMMON.SlideComp[];
}

const initState: State = {
  logsMap: {},
  slidePanelComps: [],
};

const defaultLogPeriod = 3000;

const common = createStore({
  name: 'common',
  state: initState,
  effects: {
    async uploadFile({ call }, file: FormData) {
      return call(uploadFile, file);
    },
    async getRenderPageLayout({ call }, payload: CONFIG_PAGE.RenderConfig) {
      const res = await call(getRenderPageLayout, payload);
      return res;
    },
    async fetchLog({ call }, { logKey, ...query }: { [k: string]: any; logKey: string }) {
      let lines = [] as COMMON.LogItem[];
      if (has(query, 'fetchApi') && !query.fetchApi) {
        return;
      }
      try {
        const response = await call(fetchLog, query);
        lines = response.lines;
      } catch (error) {
        // eslint-disable-next-line no-console
        console.error(error);
        lines = [];
      }
      const meta = { logKey, ...query };
      if (Array.isArray(lines)) {
        common.reducers.queryLogSuccess(lines, meta);
      }
    },
  },
  reducers: {
    queryLogSuccess(state, lines: COMMON.LogItem[], meta: { logKey: string; count?: number }) {
      const { logKey, count = 0 } = meta;
      const { logsMap } = state;
      const oldLog = logsMap[logKey] || {};
      let { emptyTimes = 0, fetchPeriod = defaultLogPeriod } = oldLog;
      const oldLines = oldLog.content || [];
      let newLines: COMMON.LogItem[] = [];
      if (lines.length) {
        emptyTimes = 0;
        fetchPeriod = defaultLogPeriod;
        newLines = count > 0 ? oldLines.concat(lines) : lines.concat(oldLines);
      } else {
        newLines = oldLines;
        emptyTimes += 1;
      }
      // 空的次数大于2次后，开始延长拉取周期，最长1分钟
      if (emptyTimes > 2 && fetchPeriod < 60 * 1000) {
        fetchPeriod += defaultLogPeriod;
      }
      state.logsMap[logKey] = { content: newLines, emptyTimes, fetchPeriod };
    },
    clearLog(state, logKey?: string) {
      if (logKey) {
        state.logsMap[logKey] = { content: [], emptyTimes: 0, fetchPeriod: defaultLogPeriod };
      } else {
        state.logsMap = {};
      }
    },
    pushSlideComp(state, payload: COMMON.SlideComp) {
      state.slidePanelComps.push(payload);
    },
    popSlideComp(state) {
      state.slidePanelComps.pop();
    },
    clearSlideComp(state) {
      state.slidePanelComps = [];
    },
  },
});

export default common;
