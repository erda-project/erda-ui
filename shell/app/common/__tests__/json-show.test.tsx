// Copyright (c) 2021 Terminus, Inc.
//
// This program is free software: you can use, redistribute, and/or modify
// it under the terms of the GNU Affero General Public License, version 3
// or later ("AGPL"), as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

/* eslint-disable no-undef */
import * as React from 'react';
import { JsonShow } from 'common';

const json = {
  a: {
    b: 'hello',
  },
};
describe('JsonShow', () => {
  it('render with data', () => {
    const wrapper = shallow(<JsonShow data={null} />);
    expect(wrapper.find('pre')).toHaveText('');
    expect(wrapper.children().find('span').at(0)).toHaveText('JSON');

    wrapper.setProps({ data: json, name: 'test' });
    expect(wrapper.find('pre')).toHaveText(JSON.stringify(json, null, 2));
    expect(wrapper.children().find('span').at(0)).toHaveText('test');
  });

  it('toggle visible by click button or switch', () => {
    const wrapper = mount(<JsonShow data={json} />);

    // const detailDom = wrapper.find('.json-detail');
    expect(wrapper.find('.json-detail')).not.toHaveClassName('slide-in');
    wrapper.find('.ant-switch').first().simulate('click');
    expect(wrapper.find('.json-detail')).toHaveClassName('slide-in');

    wrapper.find('.json-detail-btn').last().simulate('click');
    expect(wrapper.find('.json-detail')).not.toHaveClassName('slide-in');
  });
});

describe('JsonShow Snapshot', () => {
  test('renders', () => {
    const component = renderer.create(
      <JsonShow data={json} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
