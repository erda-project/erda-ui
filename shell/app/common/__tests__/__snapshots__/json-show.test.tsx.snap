// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsonShow Snapshot renders 1`] = `
<div>
  <span
    style={
      Object {
        "fontWeight": "bold",
        "verticalAlign": "middle",
      }
    }
  >
    JSON
  </span>
   
  <button
    aria-checked={false}
    checked={false}
    className="ant-switch"
    defaultChecked={false}
    onChange={[Function]}
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseUp={[Function]}
    role="switch"
    type="button"
  >
    <span
      className="ant-switch-inner"
    />
  </button>
  <div
    className="json-detail "
  >
    <button
      className="ant-btn json-detail-btn for-copy ant-btn-circle ant-btn-icon-only"
      onClick={[Function]}
      type="button"
    >
      <i
        className="anticon anticon-copy"
      >
        <svg
          aria-hidden="true"
          className=""
          data-icon="copy"
          fill="currentColor"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M832 64H296c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h496v688c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8V96c0-17.7-14.3-32-32-32zM704 192H192c-17.7 0-32 14.3-32 32v530.7c0 8.5 3.4 16.6 9.4 22.6l173.3 173.3c2.2 2.2 4.7 4 7.4 5.5v1.9h4.2c3.5 1.3 7.2 2 11 2H704c17.7 0 32-14.3 32-32V224c0-17.7-14.3-32-32-32zM350 856.2L263.9 770H350v86.2zM664 888H414V746c0-22.1-17.9-40-40-40H232V264h432v624z"
          />
        </svg>
      </i>
    </button>
    <button
      className="ant-btn json-detail-btn ant-btn-circle ant-btn-icon-only"
      onClick={[Function]}
      type="button"
    >
      <i
        className="anticon anticon-close-circle-o"
      >
        <svg
          aria-hidden="true"
          className=""
          data-icon="close-circle"
          fill="currentColor"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M685.4 354.8c0-4.4-3.6-8-8-8l-66 .3L512 465.6l-99.3-118.4-66.1-.3c-4.4 0-8 3.5-8 8 0 1.9.7 3.7 1.9 5.2l130.1 155L340.5 670a8.32 8.32 0 0 0-1.9 5.2c0 4.4 3.6 8 8 8l66.1-.3L512 564.4l99.3 118.4 66 .3c4.4 0 8-3.5 8-8 0-1.9-.7-3.7-1.9-5.2L553.5 515l130.1-155c1.2-1.4 1.8-3.3 1.8-5.2z"
          />
          <path
            d="M512 65C264.6 65 64 265.6 64 513s200.6 448 448 448 448-200.6 448-448S759.4 65 512 65zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
          />
        </svg>
      </i>
    </button>
    <pre>
      {
  "a": {
    "b": "hello"
  }
}
    </pre>
  </div>
</div>
`;
